test/e2e/storage/subpath.go

```diff
@@ -177,194 +169,462 @@ var _ = utils.SIGDescribe("Subpath", func() {
-				setInitCommand(pod, fmt.Sprintf("ln -s /bin %s", subPathDir))
+		BeforeEach(func() {
+			testInputVal := *testInput
+			if t, ok = testInputVal.(subPathTestInput); !ok {
+				framework.Failf("Fail to convert testInput to subPathTestInput")
+			}
+		})
 
-				// Pod should fail
-				testPodFailSubpath(f, pod)
-			})
+		testSubPath(&t)
+	})
+}
 
-			It("should fail if subpath file is outside the volume [Slow]", func() {
-				// Create the subpath outside the volume
-				setInitCommand(pod, fmt.Sprintf("ln -s /bin/sh %s", subPathDir))
+type subPathTestResource struct {
+	driver      testDriver
+	volType     string
+	volSource   *v1.VolumeSource
+	pvc         *v1.PersistentVolumeClaim
+	pv          *v1.PersistentVolume
+	roVolSource *v1.VolumeSource
+	sc          *storagev1.StorageClass
+	pod         *v1.Pod
+}
 
-				// Pod should fail
-				testPodFailSubpath(f, pod)
-			})
+func (s *subPathTestResource) setup(driver testDriver, testPattern testPattern) {
+	s.driver = driver
+	driverInfo := s.driver.getDriverInfo()
+	f := driverInfo.f
+	cs := f.ClientSet
+	fsType := testPattern.fsType
+
+	skipCreatingResourceForProvider(testPattern)
+
+	if testPattern.testVolType == inlineVolume {
+		framework.Logf("Creating resouce for inline volume")
+		if inlineVolumeTestDriver, ok := driver.(inlineVolumeTestDriver); ok {
+			s.volType = driverInfo.name
+			s.volSource = inlineVolumeTestDriver.getVolumeSource(false, fsType)
+			s.roVolSource = inlineVolumeTestDriver.getVolumeSource(true, fsType)
+		}
+	} else if testPattern.testVolType == staticPV {
+		framework.Logf("Creating resouce for static PV")
+		if staticPVTestDriver, ok := driver.(staticPVTestDriver); ok {
+			s.volType = fmt.Sprintf("%sPVC", driverInfo.name)
+			pvSource := staticPVTestDriver.getPersistentVolumeSource(false, fsType)
+			if pvSource != nil {
+				s.volSource, s.pv, s.pvc = CreateVolumeSourceWithPVCPV(f, driverInfo.name, pvSource, false)
+				s.roVolSource = &v1.VolumeSource{
+					PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+						ClaimName: s.pvc.Name,
+						ReadOnly:  true,
+					},
+				}
+			}
+		}
+	} else if testPattern.testVolType == dynamicPV {
+		framework.Logf("Creating resouce for dynamic PV")
+		if dynamicPVTestDriver, ok := driver.(dynamicPVTestDriver); ok {
+			claimSize := "2Gi"
+			s.volType = fmt.Sprintf("%sDynamicPVC", driverInfo.name)
+			s.sc = dynamicPVTestDriver.getDynamicProvisionStorageClass(fsType)
+
+			By("creating a StorageClass " + s.sc.Name)
+			var err error
+			s.sc, err = cs.StorageV1().StorageClasses().Create(s.sc)
+			Expect(err).NotTo(HaveOccurred())
+
+			if s.sc != nil {
+				framework.Logf("SC: %v", s.sc)
+				s.volSource, s.pv, s.pvc = CreateVolumeSourceWithPVCPVFromDynamicProvisionSC(
+					f, driverInfo.name, claimSize, s.sc, false, nil)
+				s.roVolSource = &v1.VolumeSource{
+					PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+						ClaimName: s.pvc.Name,
+						ReadOnly:  true,
```

Comment:
The readonly subpath tests may need to be restructured a little bit.  The problem with readOnly (especially with dynamic provisioning), is that you need to create the volume first, mount it writable so that it will format a fs on it, write some data to it, and then create a new pod and change it to read only.

Reply:
I think that the [latest commit](https://github.com/kubernetes/kubernetes/pull/66577/commits/2752097b35fe444f6545cb9aec18c6777365150a) is doing it and I hope that it works well.
(NFS doesn't need formatting, so its test seems working well without this code, which made me confused.)